Add new join types to KSQL

apurvam opened the pull request:
### Description 

This patch adds `LEFT`, `OUTER`, and `INNER` stream-stream and table-table joins. Additionally it adds an `INNER` stream-table join.

Majority of the logic is in `JoinNode` where a `Joiner` class hierarchy is introduced to do the heavy lifting of executing the right join. This hierarchy maximizes code reuse across the join types.

Lot's of redundant grammar constructs were removed as part of this patch. In particular, we had lexical support far `JOIN USING` instead of `JOIN ON`. We had support for CROSS joins in the grammar. These have been removed. 

Fixes #731. Fixes #760 . Fixes #1380 .

### Testing done 
1. Parser testing is done `KsqlParserTest`.  Added test cases to `SqlFormatterTest` to ensure that parsed Join nodes can be reformatted to strings correctly.
2. Correctness testing has been done using the QueryValidationTest which tests all the new join semantics.
3. Added test cases to `SchemaKStream` and `SchemaKTable` to validate that the correct KStreams APIs are being called.
4. Added test cases to `JoinNodeTest` to validate that the correct `SchemaKStream` or `SchemaKTable` methods are being called based on the join type requested.

Documentation updates will be prepared as part of a separate PR, as this is already quite large.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:
```diff
@@ -157,30 +170,39 @@ public SchemaKStream buildStream(final StreamsBuilder builder,
+                                           schemaRegistryClient,
+                                           this);
+
+    if (leftType.isStream() && rightType.isStream()) {
```
See my general comment. Perhaps we could do this with a more OO style

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:
```diff
@@ -208,36 +230,197 @@ SchemaKTable tableForJoin(
+    final int leftPartitions = left.getPartitions(kafkaTopicClient);
+    final int rightPartitions = right.getPartitions(kafkaTopicClient);
+    if (leftPartitions != rightPartitions) {
+      throw new KsqlException("Can't join " + leftType.getKqlType() + " with "
```
nit: should we use the table/stream names? Or the topic names rather than the type?

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:
```diff
@@ -208,36 +230,197 @@ SchemaKTable tableForJoin(
+                                                    joinNode.getRightKeyFieldName());
+
+      if (joinNode.spanExpression == null) {
+        throw new KsqlException("Stream-Stream joins must have a SPAN clause specified. None was "
```
nice message!

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:
```diff
@@ -208,36 +230,197 @@ SchemaKTable tableForJoin(
+      final Field field =
+          SchemaUtil
+              .getFieldByName(schema,
+                              targetKey).orElseThrow(() -> new KsqlException("couldn't find "
```
nit: I don't think `Schema` has a `toString` so this may not print anything helpful

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -229,36 +228,20 @@ public SelectValueMapper getSelectValueMapper() {
     }
   }
 
+
```
nit: extra blank line

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -229,36 +228,20 @@ public SelectValueMapper getSelectValueMapper() {
       final Field joinKey,
-      KsqlTopicSerDe joinSerDe,
-      KsqlConfig ksqlConfig
+      Serde<GenericRow> leftValueSerDe
```
nit: `final`

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -272,6 +255,116 @@ public SchemaKStream leftJoin(
+        joinSchema,
+        joinStream,
+        joinKey,
+        Arrays.asList(this, otherSchemaKStream),
```
nit: should we make this an ImmutableList?

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -272,6 +255,116 @@ public SchemaKStream leftJoin(
+      final SchemaKTable schemaKTable,
+      final Schema joinSchema,
+      final Field joinKey,
+      Serde<GenericRow> joinSerDe
```
nit: `final`

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -272,6 +255,116 @@ public SchemaKStream leftJoin(
+        joinSchema,
+        joinStream,
+        joinKey,
+        Arrays.asList(this, otherSchemaKStream),
```
ditto above

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -272,6 +255,116 @@ public SchemaKStream leftJoin(
+        joinSchema,
+        joinStream,
+        joinKey,
+        Arrays.asList(this, otherSchemaKStream),
```
ditto

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -438,4 +531,40 @@ public FunctionRegistry getFunctionRegistry() {
     return schemaRegistryClient;
   }
+
+  protected class KsqlValueJoiner implements ValueJoiner<GenericRow, GenericRow, GenericRow> {
```
`static` ?

dguy replied:
unit tests?

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java:
```diff
@@ -192,4 +190,80 @@ public SchemaKGroupedStream groupBy(
+      final Field joinKey
+  ) {
+
+    KTable joinedKTable =
```
nit: `final`

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java:
```diff
@@ -192,4 +190,80 @@ public SchemaKGroupedStream groupBy(
+        joinSchema,
+        joinedKTable,
+        joinKey,
+        Arrays.asList(this, schemaKTable),
```
ditto above and elsewhere in here

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java:
```diff
@@ -192,4 +190,80 @@ public SchemaKGroupedStream groupBy(
+      final Field joinKey
+  ) {
+
+    KTable joinedKTable =
```
nit: `final` and elsewhere

dguy commented on ksql-engine/src/test/resources/query-validation-tests/joins.json:
```diff
@@ -9,7 +9,167 @@
+      "statements": [
+        "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE bigint) WITH (kafka_topic='left_topic', value_format='DELIMITED', key='ID');",
+        "CREATE STREAM TEST_STREAM (ID bigint, F1 varchar, F2 bigint) WITH (kafka_topic='right_topic', value_format='DELIMITED', key='ID');",
+        "CREATE STREAM LEFT_OUTER_JOIN as SELECT t.id, name, value, f1, f2 FROM test t left join TEST_STREAM tt on t.id = tt.id SPAN 11 seconds;"
```
Should we also add another test that has a SPAN with BEFORE and AFTER set?

apurvam replied:
Yes. I intend to do this.

apurvam replied:
This has been added.

dguy commented on ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:
```diff
@@ -173,24 +173,27 @@ selectItem
 
 relation
-    : left=relation
-      ( CROSS JOIN right=aliasedRelation
```
nice getting rid of stuff we don't use!

dguy commented on ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java:
```diff
@@ -547,6 +547,30 @@ public Node visitSessionWindowExpression(SqlBaseParser.SessionWindowExpressionCo
     );
   }
 
+  public Node visitSpanExpression(SqlBaseParser.SpanExpressionContext ctx) {
```
nit: `final`

dguy commented on ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java:
```diff
@@ -784,52 +808,32 @@ public Node visitLogicalBinary(SqlBaseParser.LogicalBinaryContext context) {
-        throw new IllegalArgumentException("Unsupported join criteria");
-      }
+    if (context.joinCriteria().ON() == null) {
+      throw new KsqlException("Invalid join criteria specified. KSQL only supports joining on "
```
i wonder if this message should be changes to say that it only supports joins on key columns?

apurvam replied:
For streams, we repartition so that we join on the desired column. But we don't repartition for tables. I'll update this text to reflect this.

dguy commented on ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java:
```diff
@@ -784,52 +808,32 @@ public Node visitLogicalBinary(SqlBaseParser.LogicalBinaryContext context) {
+      spanExpression = (SpanExpression) visitSpanExpression(
+          context.joinWindow().spanExpression());
+    }
+    Relation left = (Relation) visit(context.left);
```
Should we add another subclass of `Relation` for `Stream`? I believe these will be two instances of `Table` at the moment which is a bit confusing later on

apurvam replied:
Hmm. In this context, the relation can either be `AliasedRelation`, which is a stream or table, or `JoinRelation` which is a join of two streams or tables. This is defined in the grammar. 

The sub relations of a join should always be an `AliasedRelation`. This is also enforced in the grammar. I updated the code to check that the actual type of `left` and `right` is `AliasedRelation`. 

Further, I synced with @hjafarpour , and the `Table` in the grammar is basically a data source: a `Table` at the grammar level could wind up being a `Stream`. So all `AliasedRelations` start as tables and then may become streams.

dguy replied:
ok, leave it as is for now. i still find it confusing though!

dguy commented on ksql-parser/src/main/java/io/confluent/ksql/parser/SqlFormatter.java:
```diff
@@ -242,45 +241,25 @@ protected Void visitTable(Table node, Integer indent) {
-        builder.append(")");
-      }
+      JoinCriteria criteria = node.getCriteria().orElseThrow(() ->
+                                                                 new KsqlException("Join "
```
nit: formatting

Would be easier to read if it was aligned left

dguy commented on ksql-parser/src/main/java/io/confluent/ksql/parser/tree/Join.java:
```diff
@@ -26,34 +26,34 @@
   public Join(NodeLocation location, Type type, Relation left, Relation right,
-              Optional<JoinCriteria> criteria) {
-    this(Optional.of(location), type, left, right, criteria);
+              Optional<JoinCriteria> criteria, SpanExpression spanExpression) {
```
maybe `SpanExpression` should be `Optional`?
Also `final`

dguy commented on ksql-parser/src/main/java/io/confluent/ksql/parser/tree/Join.java:
```diff
@@ -77,6 +77,11 @@ public Relation getRight() {
   }
 
+  public Optional<SpanExpression> getSpanExpression() {
+    return spanExpression == null
```
if it was `Optional` we don't need this. Otherwise `Optional.ofNullable(...)`

dguy commented on ksql-parser/src/test/java/io/confluent/ksql/parser/tree/SpanExpressionTest.java:
```diff
@@ -16,26 +16,24 @@
-  public int hashCode() {
-    return 0;
+  @Test
+  public void shouldDisplayCorrectStringWithSingleSpan() {
```
Wonder if we should create the `JoinWindows` in the ctor of `SpanExpression` - it would then throw an exception early on if the `JoinWindows` are invalid?

apurvam replied:
Good suggestion. done.

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKTable.java:
```diff
@@ -192,4 +190,80 @@ public SchemaKGroupedStream groupBy(
   }
+
+  @SuppressWarnings("unchecked")
+  public SchemaKTable join(
```
should we add unit tests for these new methods? Same in `SchemaKStream`

apurvam replied:
These have been added.

apurvam commented:
Thanks for the review @dguy . I refactored the `JoinNode` code to be more OO. I did not include the `JoinType` as a parameter to the factory, because I am not sure it makes sense to branch at that high level. As it stands I introduced `StreamToStreamJoiner`, `StreamToTableJoiner`, and `TableToTableJoiner`, each of which branch to do the right `LEFT`, `INNER`, or `OUTER` joins. 

I think this is an improvement over the previous iteration (thanks for the suggestion!), but further breaking out `StreamToStreamLeftJoiner`, etc. seems like it has less upside. Let me know what you think.

I also tackled most of your other comments (acknowledged with +1 on the PR). I am going to add the tests and look into the other comments now.

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:
```diff
@@ -252,7 +238,55 @@ private void ensureMatchingPartitionCounts(final KafkaTopicClient kafkaTopicClie
+    public Joiner getJoiner(DataSource.DataSourceType leftType,
+                            DataSource.DataSourceType rightType) {
+
+      if (leftType.isStream() && rightType.isStream()) {
```
i'm not that fussed, but could remove this if block if we had a `Map<Pair<DataSource.DataSourceType,DataSource.DataSourceType>, Supplier<Joiner>>` You could build it in the ctor and then this method could become something like:
```
joinerMap.getOrDefault(new Pair<>(leftType, rightType), () -> throw new KsqlException(""));
```

apurvam replied:
Thanks @dguy , I like your suggestion. I have incorporated it! The code looks much nicer now, thanks!

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:
```diff
@@ -270,7 +304,86 @@ private void ensureMatchingPartitionCounts(final KafkaTopicClient kafkaTopicClie
+    }
+
+    protected SchemaKStream maybeRePartitionByKey(final SchemaKStream stream,
+                                                final String targetKey) {
```
nit: formatting looks like it might be out, i.e., params don't line up

dguy commented on ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:
```diff
@@ -270,7 +304,86 @@ private void ensureMatchingPartitionCounts(final KafkaTopicClient kafkaTopicClie
+
+    protected SchemaKTable buildTable(final PlanNode node) {
+
+      Map<String, Object> joinTableProps = new HashMap<>(props);
```
nit: `final`

hjafarpour commented on ksql-parser/src/main/java/io/confluent/ksql/parser/SqlFormatter.java:
```diff
@@ -242,45 +241,24 @@ protected Void visitTable(Table node, Integer indent) {
+          .append(ExpressionFormatter.formatExpression(on.getExpression()))
+          .append(")");
+
+      builder.append(")");
```
Join window is missing here.

hjafarpour commented on ksql-parser/src/main/java/io/confluent/ksql/parser/tree/SpanExpression.java:
```diff
@@ -0,0 +1,101 @@
+  private final JoinWindows joinWindows;
+
+  public SpanExpression(final long before, final long after, final TimeUnit timeUnit) {
+    this(Optional.empty(), before, after, timeUnit);
```
We need two `timeUnits`, one for before and one for after.

hjafarpour commented on ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java:
```diff
@@ -717,4 +721,134 @@ public void shouldSetShowDescriptionsForShowQueriesDescriptions() {
+
+  @Test
+  public void shouldSetSpanExpressionWithSingleSpan() {
+    String statementString = "CREATE STREAM foobar as SELECT * from TEST1 JOIN ORDERS ON "
```
nit: final

hjafarpour commented on ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java:
```diff
@@ -717,4 +721,134 @@ public void shouldSetShowDescriptionsForShowQueriesDescriptions() {
+    String statementString = "CREATE STREAM foobar as SELECT * from TEST1 JOIN ORDERS ON "
+                             + "TEST1.col1 = ORDERS.ORDERID SPAN 10 SECONDS;";
+
+    Statement statement = KSQL_PARSER.buildAst(statementString, metaStore).get(0);
```
final

hjafarpour commented on ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java:
```diff
@@ -717,4 +721,134 @@ public void shouldSetShowDescriptionsForShowQueriesDescriptions() {
+
+    assertThat(statement, instanceOf(CreateStreamAsSelect.class));
+
+    CreateStreamAsSelect createStreamAsSelect = (CreateStreamAsSelect) statement;
```
final

hjafarpour commented on ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java:
```diff
@@ -717,4 +721,134 @@ public void shouldSetShowDescriptionsForShowQueriesDescriptions() {
+
+    assertThat(specification.getFrom(), instanceOf(Join.class));
+
+    Join join = (Join) specification.getFrom();
```
final, this one and other places.

hjafarpour commented on ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java:
```diff
@@ -717,4 +721,134 @@ public void shouldSetShowDescriptionsForShowQueriesDescriptions() {
+    assertThat(createStreamAsSelect.getQuery().getQueryBody(),
+               instanceOf(QuerySpecification.class));
+
+    QuerySpecification specification =
```
final

hjafarpour commented on ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java:
```diff
@@ -717,4 +721,134 @@ public void shouldSetShowDescriptionsForShowQueriesDescriptions() {
+
+  @Test
+  public void shouldSetSpanExpressionWithBeforeAndAfter() {
+    String statementString = "CREATE STREAM foobar as SELECT * from TEST1 JOIN ORDERS ON "
```
ditto final. here and the rest of the method.

hjafarpour commented on ksql-parser/src/test/java/io/confluent/ksql/parser/tree/SpanExpressionTest.java:
```diff
@@ -16,26 +16,24 @@
-    return 0;
+  @Test
+  public void shouldDisplayCorrectStringWithSingleSpan() {
+    SpanExpression expression = new SpanExpression(20, 20, TimeUnit.SECONDS);
```
final

hjafarpour commented on ksql-parser/src/test/java/io/confluent/ksql/parser/tree/SpanExpressionTest.java:
```diff
@@ -16,26 +16,24 @@
-    return toStringHelper(this).toString();
+  @Test
+  public void shouldDisplayCorrectSTringWithBeforeAndAfter() {
+    SpanExpression expression = new SpanExpression(30, 40, TimeUnit.MINUTES);
```
final

hjafarpour commented on ksql-engine/src/test/java/io/confluent/ksql/planner/plan/JoinNodeTest.java:
```diff
@@ -186,15 +189,17 @@ protected int getPartitions(KafkaTopicClient kafkaTopicClient) {
   }
 
-  @Test
+  @Test @Ignore
```
Any reason for ignoring this test?

apurvam replied:
Yes. It started to fail. I have to investigate, though I think the existing tests in this class are superseded be the new test cases I added using real mocks.

apurvam replied:
The reason both the tests were failing was because in the old code, the table was built before the stream. The new code builds the stream before the table, causing the assumptions to fail.

Since the ordering really doesn't matter in any way, I just switched the order in which the stream and table are built in a stream-table left join. This causes the tests to pass as is.

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java:
```diff
@@ -260,4 +295,252 @@ public void testGroupByMultipleColumns() {
+  @SuppressWarnings("unchecked")
+  @Test
+  public void shouldPerformStreamToStreamLeftJoin() {
+    final SchemaKStream secondSchemaKStream
```
I think this is the same in all `StreamToStreamXxxxJoin()` tests. Perhaps refactor to a field or a method

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java:
```diff
@@ -260,4 +295,252 @@ public void testGroupByMultipleColumns() {
+        .andReturn(niceMock(KStream.class));
+
+    replay(mockKstream);
+    initialSchemaKStream = new SchemaKStream(ksqlStream.getSchema(), mockKstream,
```
ditto here

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java:
```diff
@@ -260,4 +295,252 @@ public void testGroupByMultipleColumns() {
+                                 anyObject(SchemaKStream.KsqlValueJoiner.class),
+                                 eq(joinWindow),
+                                 anyObject(Joined.class)))
+        .andReturn(niceMock(KStream.class));
```
you could just return the same mock

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java:
```diff
@@ -260,4 +295,252 @@ public void testGroupByMultipleColumns() {
+  @Test
+  public void shouldPerformStreamToTableInnerJoin() {
+
+    final KStream mockKstream = EasyMock.niceMock(KStream.class);
```
`mockKstream` is used in quite a few tests, so maybe make it a field

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKStreamTest.java:
```diff
@@ -260,4 +295,252 @@ public void testGroupByMultipleColumns() {
+                 Arrays.asList(initialSchemaKStream, schemaKTable));
+  }
+
+  Schema getJoinSchema(Schema leftSchema, Schema rightSchema) {
```
nit: `final`

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKTableTest.java:
```diff
@@ -253,4 +273,136 @@ public void shouldGroupKeysCorrectly() {
+  @Test
+  public void shouldPerformTableToTableLeftJoin() {
+
+    final KTable mockKTable= EasyMock.niceMock(KTable.class);
```
field?

dguy commented on ksql-engine/src/test/java/io/confluent/ksql/structured/SchemaKTableTest.java:
```diff
@@ -253,4 +273,136 @@ public void shouldGroupKeysCorrectly() {
+        .andReturn(EasyMock.niceMock(KTable.class));
+
+    replay(mockKTable);
+    initialSchemaKTable= new SchemaKTable(ksqlTable.getSchema(), mockKTable,
```
Same comments as in the `SchemaKStreamTest` - can extract the common code in the tests to methods etc.

apurvam commented:
@hjafarpour @dguy 

Thanks for the reviews! I have addressed all the comments and finished writing tests for all the new code. Please take a look.

rodesai commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -438,4 +532,41 @@ public FunctionRegistry getFunctionRegistry() {
+    }
+
+    private void fillWithNulls(final List<Object> columns, final int numToFill) {
+      for (int i = 0; i < numToFill; ++i) {
```
Consider making this more functional by returning a new list instead of updating: ```return Arrays.asList(new Object[numToFill]);```

rodesai commented on ksql-engine/src/test/java/io/confluent/ksql/planner/plan/JoinNodeTest.java:
```diff
@@ -238,4 +302,619 @@ public void shouldHaveAllFieldsFromJoinedInputs() {
+
+    verify(left, right, leftSchemaKStream, rightSchemaKStream);
+
+    assertEquals(leftKeyFieldName, joinNode.getLeftKeyFieldName());
```
Why have all these checks? These fields are just set by the constructor.

apurvam replied:
That's the behavior now. But we want to make sure that if it changes, at least we catch it and do it consciously. No harm in checking IMO.

rodesai commented on ksql-engine/src/test/java/io/confluent/ksql/planner/plan/JoinNodeTest.java:
```diff
@@ -238,4 +302,619 @@ public void shouldHaveAllFieldsFromJoinedInputs() {
+      fail("should have raised an exception since the number of partitions on the input sources "
+           + "don't match");
+    } catch (KsqlException e) {
+      // good!
```
maybe do some validation of the message to make sure we're throwing for the right reason.

rodesai commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -438,4 +532,41 @@ public FunctionRegistry getFunctionRegistry() {
     return schemaRegistryClient;
   }
+
+  protected static class KsqlValueJoiner
```
We should try and test this class.

apurvam replied:
There is a unit test for it for each (nonNull, nonNull), (nonNull, null), and (null, nonNull) scenario.

apurvam commented on ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java:
```diff
@@ -547,6 +547,38 @@ public Node visitSessionWindowExpression(SqlBaseParser.SessionWindowExpressionCo
   }
 
+  public Node visitSpanExpression(final SqlBaseParser.SpanExpressionContext ctx) {
+    Pair<Long, TimeUnit> beforeSize;
```
can't be final since they are updated in the branches below.

rodesai replied:
They aren't updated - they're initialized in the branches.

apurvam replied:
ah. I thought `final` was like `const`. Updated.

apurvam commented on ksql-parser/src/main/java/io/confluent/ksql/parser/tree/SpanExpression.java:
```diff
@@ -0,0 +1,122 @@
+  private final TimeUnit afterTimeUnit;
+  private final JoinWindows joinWindows;
+
+  public SpanExpression(final long size, final TimeUnit timeUnit) {
```
yes.

apurvam commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -438,4 +532,41 @@ public FunctionRegistry getFunctionRegistry() {
+      if (right != null) {
+        columns.addAll(right.getColumns());
+      } else {
+        fillWithNulls(columns, rightSchema.fields().size());
```
no. It will just append nulls to the end of the array. There is a test for this.

rodesai replied:
Ah - for some reason I read the loop below as using set.

apurvam commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -234,44 +235,137 @@ public SchemaKStream leftJoin(
-            (ValueJoiner<GenericRow, GenericRow, GenericRow>) (leftGenericRow, rightGenericRow) -> {
-              List<Object> columns = new ArrayList<>(leftGenericRow.getColumns());
-              if (rightGenericRow == null) {
-                for (int i = leftGenericRow.getColumns().size();
```
No. We only supported `LEFT` joins, and as such only the right could be null. With `OUTER` join support, either `LEFT` or `RIGHT` can be null.

rodesai replied:
I meant filling in the right side of the result with nulls based on the number of columns on the left if the right is null.

rodesai replied:
Oh right. I’ve been using the new java8 apis so much I’ve forgotten how to read loops!

apurvam commented on ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:
```diff
@@ -200,44 +211,294 @@ SchemaKTable tableForJoin(
+                                                    joinNode.getRightKeyFieldName());
+
+      if (joinNode.spanExpression == null) {
+        throw new KsqlException("Stream-Stream joins must have a SPAN clause specified. None was "
```
Good call. Will add the check and a test.

hjafarpour commented on ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:
```diff
@@ -200,44 +211,294 @@ SchemaKTable tableForJoin(
-    return right.getPartitions(kafkaTopicClient);
+  private String getSourceName(PlanNode node) {
+    if (!(node instanceof StructuredDataSourceNode)) {
+      throw new KsqlException("The source for a join must be a Stream or a Table.");
```
Is this needed? Seems that this method only is called for creating an exception message and if you have this exception, it would supersede the other exception message!

apurvam replied:
The only reason for this method is support printing a nicer error message which name each source in case the number of partitions don't match. I changed the exception to be an instance of `RunTimeException`, since it is just a sanity check and should never be raised unless there is bug.

hjafarpour commented on ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:
```diff
@@ -115,7 +115,7 @@ querySpecification
 windowExpression
     : (IDENTIFIER)?
-     ( tumblingWindowExpression | hoppingWindowExpression | sessionWindowExpression)
+     ( tumblingWindowExpression | hoppingWindowExpression | sessionWindowExpression )
```
extra space?

apurvam replied:
made it symmetric on either side, and also consistent with other rules.

hjafarpour commented on ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:
```diff
@@ -173,24 +173,31 @@ selectItem
-      | NATURAL joinType JOIN right=aliasedRelation
-      )                                           #joinRelation
-    | aliasedRelation                             #relationDefault
+    : left=aliasedRelation joinType JOIN right=aliasedRelation joinCriteria joinWindow #joinRelation
```
Join window is optional. Use `joinWindow?`

hjafarpour commented on ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:
```diff
@@ -173,24 +173,31 @@ selectItem
-    | LEFT OUTER?
-    | RIGHT OUTER?
-    | FULL OUTER?
+    | OUTER?
```
Change it back to `FULL OUTER?`.
Syntax should be the same as standard: https://www.w3schools.com/sql/sql_join_full.asp

hjafarpour commented on ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:
```diff
@@ -173,24 +173,31 @@ selectItem
-    | RIGHT OUTER?
-    | FULL OUTER?
+    | OUTER?
+    | LEFT?
```
Change back to `LEFT OUTER?`.
Standard SQL syntax: https://www.w3schools.com/sql/sql_join_left.asp

apurvam commented on ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java:
```diff
@@ -234,44 +235,137 @@ public SchemaKStream leftJoin(
+
+  @SuppressWarnings("unchecked")
+  public SchemaKStream leftJoin(
+      final SchemaKStream otherSchemaKStream,
```
Not sure that `rightSchemaKStream` makes anything much clearer since the left side is just `this`, and not `leftSchemaKStream`.

hjafarpour commented on ksql-parser/src/main/java/io/confluent/ksql/parser/SqlFormatter.java:
```diff
@@ -242,45 +241,26 @@ protected Void visitTable(Table node, Integer indent) {
+
+      builder.append(")");
+
+      node.getSpanExpression().map((e) -> builder.append(' ').append(e.toString()));
```
Tests for join cases?

apurvam commented on ksql-engine/src/test/java/io/confluent/ksql/QueryTranslationTest.java:
```diff
@@ -265,7 +267,7 @@ public QueryTranslationTest(final String name, final Query query) {
       String test;
       while ((test = reader.readLine()) != null) {
-        if (test.endsWith(".json")) {
+        if (test.endsWith("json")) {
```
actually, no. I updated it to just run `joins.json` while I was developing the code, and then forget to put the `.` back.

apurvam commented on ksql-engine/src/test/resources/query-validation-tests/joins.json:
```diff
@@ -9,7 +9,190 @@
+        {"topic": "right_topic", "key": 0, "value": "0,a,10", "timestamp": 15000},
+        {"topic": "right_topic", "key": 100, "value": "100,newblah,150", "timestamp": 16000},
+        {"topic": "left_topic", "key": 90, "value": "90,ninety,90", "timestamp": 17000},
+        {"topic": "left_topic", "key": 0, "value": "0,bar,99", "timestamp": 30000}
```
I added a test case to show case this.

ppadovani commented:
Are there any docs that describe the behavior of WITHIN/SPAN? All I have been able to find are examples of the syntax in the test cases.

apurvam commented:
@ppadovani the new syntax is documented here: https://docs.confluent.io/current/ksql/docs/syntax-reference.html#create-stream-as-select

As for a description of the window semantics, the best we have right now are the kafka streams documentation: https://docs.confluent.io/current/streams/developer-guide/dsl-api.html#kstream-kstream-join

The semantics of the `WITHIN` clause in a stream-stream KSQL join is the same as the semantics of `JoinWindows` in KStreams.
